sr_utilities_common/src/sr_utilities_common/aws_manager.py

```diff
@@ -71,66 +68,189 @@ def get_bucket_structure_with_prefix(self, bucket_name, prefix):
+        for file_full_path, aws_path in zip(self.file_full_paths, self.aws_paths):
+            directory = os.path.dirname(file_full_path)
+            if not os.path.exists(directory):
+                os.makedirs(directory)
             try:
-                self._client.download_file(bucket_name, self.aws_path, self.file_full_path)
+                self._client.download_file(bucket_name, aws_path, file_full_path)
                 downloadSucceded = True
             except Exception as e:
                 rospy.logwarn("File download failed. " + str(e))
         return downloadSucceded
 
-    def upload(self, bucket_name, files_base_path, files_folder_path, file_names):
-        self._prepare_structure(bucket_name, files_base_path, files_folder_path, file_names)
+    def upload(self, bucket_name, files_base_path, files_folder_path, file_names, bucket_subfolder):
+        self._prepare_structure_upload(files_base_path, files_folder_path, file_names, bucket_subfolder)
         uploadSucceded = False
-        for self.file_full_path, self.aws_path in zip(self.file_full_paths, self.aws_paths):
+        for file_full_path, aws_path in zip(self.file_full_paths, self.aws_paths):
             try:
-                self._client.upload_file(self.file_full_path, bucket_name, self.aws_path)
+                self._client.upload_file(file_full_path, bucket_name, aws_path)
                 uploadSucceded = True
             except Exception as e:
                 rospy.loginfo("File upload failed" + str(e))
         return uploadSucceded
 
 
+def gather_all_files_local(files_base_path, files_folder_path):
+    path_string = f"{files_base_path}/{files_folder_path}"
+    file_names = []
+    for path, _, files in os.walk(path_string):
+        for file_name in files:
+            # We only want the file name, not bucket subfolder.
+            file_names.append(f"{path}/{file_name}"[len(path_string)+1:])
+    return file_names
+
+
+def validated_files_to_be_downloaded(bucket_name, files_base_path, files_folder_path,
+                                     file_names, bucket_subfolder):
+    print_msg = f"\nFrom bucket {bucket_name} downloading the files:"
+    print_msg_2 = ""
+    for file in file_names:
+        print_msg_2 += f"\n    {files_base_path}/{files_folder_path}/{file}"
+        if bucket_subfolder:
+            print_msg += f"\n    {bucket_subfolder}/{file}"
+        else:
+            print_msg += f"\n    {file}"
+    print_msg += f"\n\nThese files will be downloaded too {files_base_path}/{files_folder_path}:"
+    print_msg += print_msg_2
+    rospy.loginfo(print_msg)
+    value = input("Would you like to download these files? (Y/N) ")
+    value = value.lower().strip()
+    if value == "y":
+        return True
+    elif value == "n":
+        return False
+    else:
+        rospy.logerr("Select a valid option")
+        rospy.signal_shutdown("")
+        exit(1)
```

Comment:
Same for below: 
Maybe add a small loop, let's say repeat the y/n question 3 times and then shutdown. Now if it fails, you'll have to relaunch all. Or add another option to abort?